packages/contract-wrappers/src/fetchers/asset_balance_and_proxy_allowance_fetcher.ts

```diff
@@ -18,60 +19,101 @@ export class AssetBalanceAndProxyAllowanceFetcher implements AbstractBalanceAndP
     }
     public async getBalanceAsync(assetData: string, userAddress: string): Promise<BigNumber> {
         const decodedAssetData = assetDataUtils.decodeAssetDataOrThrow(assetData);
-        if (decodedAssetData.assetProxyId === AssetProxyId.ERC20) {
-            const decodedERC20AssetData = decodedAssetData as ERC20AssetData;
-            const balance = await this._erc20Token.getBalanceAsync(decodedERC20AssetData.tokenAddress, userAddress, {
-                defaultBlock: this._stateLayer,
-            });
-            return balance;
-        } else {
-            const decodedERC721AssetData = decodedAssetData as ERC721AssetData;
-            const tokenOwner = await this._erc721Token.getOwnerOfAsync(
-                decodedERC721AssetData.tokenAddress,
-                decodedERC721AssetData.tokenId,
-                {
+        let balance: BigNumber | undefined;
+        switch (decodedAssetData.assetProxyId) {
```

Comment:
I see this `switch (decodedAssetData.assetProxyId)` pattern come up a lot. We could optionally use ["type guards"](https://www.typescriptlang.org/docs/handbook/advanced-types.html) to make the types a little easier to manage and avoid type assertions.

For example:

```ts
function isERC20AssetData(assetData: SingleAssetData | MultiAssetData): assetData is ERC20AssetData {
    return assetData.assetProxyId === AssetProxyId.ERC20;
}
```

If we write this type guard, the TypeScript compiler knows that when `isERC20AssetData` returns true the type should be `ERC20AssetData`.